doc/dev/osd_internals/multi_object_write_transaction.rst

```diff
@@ -0,0 +1,117 @@
+==============================
+Multi-Object Write Transaction
+==============================
+
+Multi-object write transaction enables a group of write operations 
+on multi-objects complete atomically.
+
+Description of the Process
+--------------------------
+
+The high level process is,
+
+  1. Define the following class,
+
+     class *CEPH_RADOS_API MultiObjectWriteOperation*
+
+     {
+
+       std::map<std::string, ObjectWriteOperation*> slaves;
+
+       ObjectWriteOperation master;
+
+     };
+
+     based on this, client could send a group of ops on multi-objects 
+     through *MOSDOp* to the primary osd in charge of the *master* object 
+     which is explicitly specified by client, call this osd *MASTER*
+
+  #. *MASTER* receives the multi-object write operation request, performs 
+     dead lock check (detailed below), extracts *ObjectWriteOperations* 
+     by objects from *slaves*, and sends them to corresponding primary 
+     osds individually, calls those osds *SLAVE*, the transaction 
+     enters the *PREPARE* state
+
+  #. *SLAVE* receives *MOSDOp* from *MASTER*, note each op received by *SLAVE*
+     is a group of write operations on a single slave object. *SLAVE* 
+     performs dead lock check, does the normal *prepare_transaction()* stuff, 
+     if everything is ok, *SLAVE* will lock the object, and delay further 
+     read/write operations on this object until the transaction finishs. 
+     *SLAVE* sends a *QUERY* message to *MASTER*, and returns *MASTER* 
+     *EINPROGRESS*. The *QUERY* message is used for *MASTER* to notify *SLAVE* 
+     when to release transaction context, *MASTER* will not ack *QUERY* until 
+     itself has released transaction context
```

Comment:
Yep, at PREPARE, it does not persist anything. From the aspect of correctness, PREPARE and LOG could be merged. PREPARE is kept separate for better performance.  As also mentioned by @liewegas , before doing any real transaction stuff, a lightweight in-memory only dead lock and other validity check is desirable, if we do LOG first for all SLAVEs, the roll back overhead will be much higher.

If we agree to keep PREPARE alone, then after PREPARE, any slave must immediately lock the object (in memory) to prevent the object being modified by further write operations, before the transaction finishes. If the SLAVE undergoes an interval change after PREPARE, yes, the new SLAVE will forget the lock and transaction, the MASTER will send LOG/resend PREPARE to the new SLAVE, now, there are two cases,
(1) For the first case, there is no new transaction intervening, then the SLAVE will respond MASTER EPERM to indicate that it does not know this transaction at all, then MASTER will resend PREPARE to this SLAVE, and the SLAVE will redo validity check, lock the object, proceed as before;
(2)	For the second case, there is other transaction intervening, then the SLAVE will find there is other pending transaction on this object, it will fail the dead lock check, and respond MASTER EDEADLK, then MASTER will notify other SLAVEs to roll back.
In either of the cases, it will not cause problem and the semantics of transaction will not be violated. Since the real object modification happens only after MASTER has collected all LOG success message from SLAVES, at that time, all the lock have been persisted.